client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala

```diff
@@ -719,6 +780,104 @@ class DeltaSharingRestClient(
+    val schemaName = URLEncoder.encode(table.schema, "UTF-8")
+    val tableName = URLEncoder.encode(table.name, "UTF-8")
+    val encodedQueryId = URLEncoder.encode(queryId, "UTF-8")
+    val target = getTargetUrl(
+      s"/shares/$shareName/schemas/$schemaName/tables/$tableName/queries/$encodedQueryId")
+    target
+  }
+
+  private def getTableQueryInfo(
+      table: Table,
+      queryId: String,
+      maxFiles: Option[Int],
+      pageToken: Option[String]): (Long, String, Seq[String]) = {
+    val target: String = getQueryInfoTargetUrl(table, queryId)
+    val request: GetQueryTableInfoRequest = GetQueryTableInfoRequest(
+      queryId = queryId,
+      maxFiles = maxFiles,
+      pageToken = pageToken)
+
+    getNDJson(target, request)
+  }
+
+  /*
+  * Check if the query is still pending. The first line of the response will
+  * be a query status object if the query is still pending.
+  * The method return (queryResultLines, queryId, queryPending) tuple. If the queryPending
+  * is false it means the query is finished and the queryResultLines contains the result.
+   */
+  private def checkQueryPending(lines: Seq[String]): (Seq[String], Option[String], Boolean) = {
+    val queryStatus = JsonUtils.fromJson[SingleAction](lines(0)).queryStatus
+
+    if (queryStatus == null) {
+      (lines, None, false)
+    } else {
+      if (queryStatus.queryId == null) {
+        throw new IllegalStateException(
+          "QueryId is not returned in the first line of the response." + lines(0))
+      }
+
+      (lines.drop(1), Some(queryStatus.queryId), true)
+    }
+  }
+
+  /*
+  * Get NDJson data with async query.
+  * This method is used when we get the result of table query
+  * If the query is async mode and still running, this method
+  * will keep polling the query id until the query is finished.
+  * and return the result back.
+  * If the query is in sync mode it will return the query result
+  * directly.
+   */
+  private def getNDJsonWithAsync(
+                table: Table,
+                target: String,
+                request: QueryTableRequest): (Long, String, Seq[String], Option[String]) = {
+    // Initial query to get NDJson data
+    val (initialVersion, initialRespondedFormat, initialLines) = getNDJson(target, request)
+
+    // Check if the query is still pending
+    var (lines, queryIdOpt, queryPending) = checkQueryPending(initialLines)
+
+    var version = initialVersion
+    var respondedFormat = initialRespondedFormat
+
+    val startTime = System.currentTimeMillis()
+    // Loop while the query is pending
+    while (queryPending) {
+      if (System.currentTimeMillis() - startTime > asyncQueryMaxDuration) {
+        throw new IllegalStateException("Query is timed out after " +
+          s"${asyncQueryMaxDuration} ms. Please try again later.")
+      }
+
+      val queryId = queryIdOpt.get
+      logInfo(s"Query is still pending. Polling queryId: ${queryId}")
```

Comment:
I worry this results in too many logging as it's one log every second.

Reply:
this basically a config value, and the default config value is 10s, I changed the default value of the constructor here also to make it consistent.